custom_enumerable_methods.rb

```diff
@@ -0,0 +1,249 @@
+    return false unless block_given? || !my_parameter.nil?
+
+    result = true
+    if self.class == Array
+      my_each do |x|
+        if block_given?
+          result = false if yield(x)
+        elsif my_parameter.class == Regexp
+          result = false if x.match(my_parameter)
+        elsif my_parameter.class <= Numeric
+          result = false if x == my_parameter
+        elsif x.class <= my_parameter
+          result = false
+        end
+        break unless result
+      end
+    else
+      my_each do |key, value|
+        result = false if yield(key, value)
+        break unless result
+      end
+    end
+    result
+  end
+
+  def my_count(my_variable = nil)
+    my_counter = 0
+    if block_given?
+
+      if self.class == Array
+        my_each do |n|
+          my_counter += 1 if yield(n)
+        end
+      else
+        my_each do |key, value|
+          my_counter += 1 if yield(key, value)
+        end
+      end
+    elsif !block_given? && my_variable.nil?
+      return length
+    elsif !block_given? && !my_variable.nil?
+      my_each do |n|
+        my_counter += 1 if n == my_variable
+      end
+    end
+    my_counter
+  end
+
+  def my_map
+    return to_enum unless block_given?
+
+    new_array = []
+    if self.class == Array
+      my_each do |n|
+        new_array << yield(n)
+      end
+    else
+      my_each do |key, value|
+        new_array << yield(key, value)
+      end
+    end
+    new_array
+  end
+
+  def my_inject(symbol = nil, start_value = nil)
+    if symbol.class != Symbol
+      holder = symbol
+      symbol = start_value
+      start_value = holder
+    end
+    initial_value = false
+    initial_value = true unless start_value.nil?
+    value = start_value || first
+    case symbol
+    when :+
+      if !initial_value
+        drop(1).my_each do |x|
+          value += x
+        end
+      else
+        my_each do |x|
+          value += x
+        end
+      end
+    when :*
+      if !initial_value
+        drop(1).my_each do |x|
+          value *= x
+        end
+      else
+        my_each do |x|
+          value *= x
+        end
+      end
+    when :/
+      if !initial_value
+        drop(1).my_each do |x|
+          value /= x
+        end
+      else
+        my_each do |x|
+          value /= x
+        end
+      end
+    when :-
+      if !initial_value
+        drop(1).my_each do |x|
+          value -= x
+        end
+      else
+        my_each do |x|
+          value -= x
+        end
+      end
+    when :**
+      if !initial_value
+        drop(1).my_each do |x|
+          value **= x
+        end
+      else
+        my_each do |x|
+          value **= x
+        end
+      end
+    else
+      if !initial_value
+        drop(1).my_each do |x|
+          value = yield(value, x)
+        end
+      else
+        my_each do |x|
+          value = yield(value, x)
+        end
+      end
+    end
+    value
+  end
+end
```

Comment:
* [ ] my_inject combines all elements of enum by applying a binary operation, specified by a block
  ```ruby
    range = Range.new(5, 50)
    p range.my_inject(4) { |prod, n| prod * n } # returns this error
    # NameError:
       undefined local variable or method `length' for 5..50:Range
     # ./custom_enumerable_methods.rb:4:in `my_each'
     # ./custom_enumerable_methods.rb:225:in `my_inject'
  ```
* [ ] my_inject when a symbol is specified combines each element of the collection by applying the symbol as a named method
  ```ruby
    range = Range.new(5, 50)
    p range.my_inject(2, :*) # returns this error
    # NameError:
       undefined local variable or method `length' for 5..50:Range
     # ./custom_enumerable_methods.rb:4:in `my_each'
     # ./custom_enumerable_methods.rb:185:in `my_inject'
  ```